### PR TITLE
[Feature] barlowtwins

### PR DESCRIPTION
**Pull request by xcnick**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

This PR implements barlow twins algorithm.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by fangyixiao18 — COMMENTED** [review]
Thank you for your contribution. Do you have any results or models of barlowtwins based on this PR?

**Comment by xcnick** [comment]
> Thank you for your contribution. Do you have any results or models of barlowtwins based on this PR?

Sorry, due to lack of computing resources, I only performed the pretraining steps on RTX2080Ti * 2, and observed loss decreasing. Therefore, the config parameters may need to be further adjusted.

I will use 4 * V100 to verify this PR later if possible.

**Comment by fangyixiao18** [comment]
> Sorry, due to lack of computing resources, I only performed the pretraining steps on RTX2080Ti * 2, and observed loss decreasing. Therefore, the config parameters may need to be further adjusted.
> 
> I will use 4 * V100 to verify this PR later if possible.

OK, can you checkout a new branch from dev_v0.7.0 and re-create a PR to the dev_v0.7.0 branch? And if you have any updated results,  you can push new commits to the new PR. 
We look forward that your PR can be merged and the BarlowTwins can be released in version 0.7.0 at the begining of the next month.

**Comment by xcnick** [comment]
> > Sorry, due to lack of computing resources, I only performed the pretraining steps on RTX2080Ti * 2, and observed loss decreasing. Therefore, the config parameters may need to be further adjusted.
> > I will use 4 * V100 to verify this PR later if possible.
> 
> OK, can you checkout a new branch from dev_v0.7.0 and re-create a PR to the dev_v0.7.0 branch? And if you have any updated results, you can push new commits to the new PR. We look forward that your PR can be merged and the BarlowTwins can be released in version 0.7.0 at the begining of the next month.

OK.

**Comment by xcnick** [comment]
Using the parameters in the code for pretraining in 4*V100, the loss value is about 2000. The top1 acc is 40% when finetune in imagenet1k. So the model did not converge.
Could you help me for the model training?

**Comment by YuanLiuuuuuu** [comment]
> Using the parameters in the code for pretraining in 4*V100, the loss value is about 2000. The top1 acc is 40% when finetune in imagenet1k. So the model did not converge. Could you help me for the model training?



> Using the parameters in the code for pretraining in 4*V100, the loss value is about 2000. The top1 acc is 40% when finetune in imagenet1k. So the model did not converge. Could you help me for the model training?

If you do not have enough computation resources, you can choose a setting, which has lower computation overhead. Under this setting, you can run the official code and get the log. Finally, you can reproduce the results, according to the log.

**Comment by YuanLiuuuuuu** [comment]
We have already created a new branch, dev_v0.8.0, for next released version. Please re-submit a PR to merge all your modifications to that branch as soon as possible. Thanks a lot !

**Comment by YuanLiuuuuuu** [comment]
> Using the parameters in the code for pretraining in 4*V100, the loss value is about 2000. The top1 acc is 40% when finetune in imagenet1k. So the model did not converge. Could you help me for the model training?

Please add the same [decorator](https://github.com/open-mmlab/mmselfsup/pull/207#issuecomment-1061464369) to your algorithm to pass CI.

**Comment by fangyixiao18** [comment]
Do you have any updated training results?

**Comment by xcnick** [comment]
> Do you have any updated training results?

Sorry, the learning rate parameter is adjusted according to the [official code](https://github.com/facebookresearch/barlowtwins/blob/main/main.py), but the loss is around 2000. In contrast, the loss is around 500 in [official log](https://dl.fbaipublicfiles.com/barlowtwins/ljng/stats.txt) when training epoch is 100.

**Comment by fangyixiao18** [comment]
> Sorry, the learning rate parameter is adjusted according to the [official code](https://github.com/facebookresearch/barlowtwins/blob/main/main.py), but the loss is around 2000. In contrast, the loss is around 500 in [official log](https://dl.fbaipublicfiles.com/barlowtwins/ljng/stats.txt) when training epoch is 100.

Could you provide your personal contact information? I have sent an e-mail to your gmail account which is displayed on your github homepage.

File: configs/selfsup/barlowtwins/barlowtwins_resnet50_8xb32-accum8-coslr-100e_in1k.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+_base_ = 'barlowtwins_resnet50_8xb256-coslr-300e_in1k.py'
+
```
Comment (xcnick) [inline configs/selfsup/barlowtwins/barlowtwins_resnet50_8xb32-accum8-coslr-100e_in1k.py:2]: May need to add  `data = dict(samples_per_gpu=32)` to modify the batch_size.

**Review by Jiahao000 — COMMENTED** [review]
Better to reorganize the algorithms in model zoo in their chronological order. Others LGTM.

File: mmselfsup/models/algorithms/barlowtwins.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        neck (dict): Config dict for module of deep features to compact
+            feature vectors. Defaults to None.
+        head (dict): Config dict for module of loss functions.
+            Defaults to None.
+        init_cfg (dict): Config dict for weight initialization.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 backbone: object,
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/barlowtwins.py]: type hint should be dict

File: mmselfsup/models/algorithms/barlowtwins.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                 neck: object = None,
+                 head: object = None,
+                 init_cfg: Optional[dict] = None,
+                 **kwargs) -> None:
+        super(BarlowTwins, self).__init__(init_cfg)
+        assert neck is not None
+        self.encoder = nn.Sequential(
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/barlowtwins.py]: This is no need to create an extra ``encoder``, just using ``backbone`` and ``neck`` is OK

File: mmselfsup/models/heads/latent_pred_head.py [hunk 4]
```diff
@@ -18,11 +18,11 @@ class LatentPredictHead(BaseModule):
         predictor (dict): Config dict for module of predictor.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/heads/latent_pred_head.py]:
```suggestion
        predictor (dict): Config dict for the predictor.
```

File: mmselfsup/models/heads/latent_pred_head.py [hunk 5]
```diff
@@ -74,3 +80,51 @@ def forward(self, input, target):
...
+
+
+@HEADS.register_module()
+class LatentCrossCorrelationHead(BaseModule):
+    """Head for latent feature cross correlation. Part of the code is borrowed
+    from:
+    `https://github.com/facebookresearch/barlowtwins/blob/main/main.py>`_.
+
+    Args:
+        in_channels (int): Number of input channels.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/heads/latent_pred_head.py:92]: missing ``lambd``

File: mmselfsup/models/heads/latent_pred_head.py [hunk 6]
```diff
@@ -74,3 +80,51 @@ def forward(self, input, target):
...
+    """
+
+    def __init__(self, in_channels: int, lambd: float = 0.0051) -> None:
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/heads/latent_pred_head.py:96]:
```suggestion
    def __init__(self, in_channels: int, lamb: float = 0.0051) -> None:
```

Comment (fangyixiao18) [reply mmselfsup/models/heads/latent_pred_head.py:96]: Is there any docs to suggest `lamb`?

File: docs/en/model_zoo.md [hunk 7]
```diff
@@ -37,28 +38,29 @@ In the following tables, we only display ImageNet linear evaluation, ImageNet fi
...
-| SimMIM       | [simmim_swin-base_16xb128-coslr-100e_in1k-192](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192.py) |         | 82.9      |
+| Algorithm | Config                                                                                                                                                                     | Remarks | Top-1 (%) |
+| --------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | --------- |
+| MAE       | [mae_vit-base-p16_8xb512-coslr-400e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mae/mae_vit-base-p16_8xb512-coslr-400e_in1k.py)              |         | 83.1      |
+| SimMIM    | [simmim_swin-base_16xb128-coslr-100e_in1k-192](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/simmim/simmim_swin-base_16xb128-coslr-100e_in1k-192.py) |         | 82.9      |
 
 ### COCO17 Object Detection and Instance Segmentation
```
Comment (Jiahao000) [inline docs/en/model_zoo.md:65]: Other benchmark tables in model_zoo should also be reordered.

**Review by Jiahao000 — CHANGES_REQUESTED** [review]

**Review by Jiahao000 — APPROVED** [review]

**Comment by YuanLiuuuuuu** [comment]
@fangyixiao18 Please update ``index.rst`` in ``docs``

File: mmselfsup/models/algorithms/barlowtwins.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            tuple[Tensor]: backbone outputs.
+        """
+        x = self.backbone(img)
+        return x
+
+    def forward_train(self, img: List[torch.Tensor]) -> dict:
+        """Forward computation during training.
+
+        Args:
+            img (list[Tensor]): A list of input images with shape
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/barlowtwins.py]: list -> List

File: mmselfsup/models/algorithms/barlowtwins.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                (N, C, H, W). Typically these should be mean centered
+                and std scaled.
+        Returns:
+            loss[str, Tensor]: A dictionary of loss components
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/algorithms/barlowtwins.py]:
```suggestion
            dict[str, Tensor]: A dictionary of loss components
```

**Review by YuanLiuuuuuu — APPROVED** [review]
